### PR TITLE
feat: add signing

### PR DESCRIPTION
**Pull request by wilfriedroset**
This pull request add the capability to pass puppet environment to the provider and  certificates signing.
On top of that the deps have been update, go version has been bumped to v1.17.

Regarding puppet environment, this is mandatory when working with puppet server < v6.
While puppet v5 is EoL it is still the default version shipped in common distribution such as Debian 11.
This feature is very handy when working in environment where you are not running puppet with autosign enabled.

Please note that this PR will need to be updated (go.mod related) once https://github.com/camptocamp/go-puppetca/pull/3 has been merged.

### CODE REVIEW

File: puppetca/resource_puppetca_certificate.go [hunk 1]
```diff
@@ -55,14 +62,36 @@ func resourcePuppetCACertificateCreate(d *schema.ResourceData, meta interface{})
...
+func findCert(client puppetca.Client, name, env string) (*string, error) {
+	cert, err := client.GetCertByName(name, env)
+
+	if err != nil || cert == "" {
+		return nil, errors.New("not found")
+	}
+	return &cert, nil
+}
+
+func SignCert(client puppetca.Client, name, env string) resource.StateRefreshFunc {
```
[the remark this answers is not in the document]
Comment (wilfriedroset) [reply puppetca/resource_puppetca_certificate.go]: No, this method does not need to be exposed.
I've amended to reflect that.

File: go.sum [hunk 2]
```diff
@@ -73,6 +73,8 @@ github.com/bgentry/speakeasy v0.1.0 h1:ByYyxL9InA1OWqxJqqp2A5pYHUrCiAL6K3J+LKSsQ
 github.com/bgentry/speakeasy v0.1.0/go.mod h1:+zsyZBPWlz7T6j88CTgSN5bM796AkVf0kBD4zp0CCIs=
 github.com/camptocamp/go-puppetca v0.0.0-20200918132719-43b6b595ceee h1:dzta/TfPpcSuq3EHeVWmu+vJL037iB9b3EMKqi9M26A=
 github.com/camptocamp/go-puppetca v0.0.0-20200918132719-43b6b595ceee/go.mod h1:5F/WXQW/rXgXl9jBlEYotd30Gn0luLxdJWTbOZxSlqQ=
```
Comment (raphink) [inline go.sum:75]: A run of `go mod tidy` should remove the sums for old lib versions.

File: go.mod [hunk 3]
```diff
@@ -2,6 +2,11 @@ module github.com/camptocamp/terraform-provider-puppetca
 
 go 1.17
 
+require (
+	github.com/camptocamp/go-puppetca v0.0.0-20220207140217-c3faf5528998
+	github.com/hashicorp/terraform-plugin-sdk v1.17.2
+)
+
```
Comment (raphink) [inline go.mod:9]: All the libs should go in the same `require` block

File: puppetca/resource_puppetca_certificate.go [hunk 4]
```diff
@@ -55,14 +75,46 @@ func resourcePuppetCACertificateCreate(d *schema.ResourceData, meta interface{})
 	return resourcePuppetCACertificateRead(d, meta)
 }
 
-func findCert(client puppetca.Client, name string) resource.StateRefreshFunc {
+func findCert(client puppetca.Client, name, env string) (string, string, error) {
+	cert, err := client.GetCertByName(name, env)
+
+	if err != nil || cert == "" {
+		if strings.Contains(err.Error(), "404 Not Found") {
```
Comment (raphink) [inline puppetca/resource_puppetca_certificate.go:82]: Does the PuppetCA always return a 200? Ideally, the code should be checked in the HTTP error, not in the body.

Comment (wilfriedroset) [reply puppetca/resource_puppetca_certificate.go:82]: Puppet returns 404 which is passed from go-puppetca and this bubble up to terraform. If we return the 404 (which is expected in this case) terraform stops and claims that the provider has a bug.
I don't know how to check the HTTP error :D but I will have a look at this.

Comment (wilfriedroset) [reply puppetca/resource_puppetca_certificate.go]: As per https://github.com/camptocamp/go-puppetca/blob/master/puppetca/puppetca.go#L184 we do not have the status code of the response. Neither the response nor the err are returned
```golang
	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
		return "", fmt.Errorf("failed to %s URL %s, got: %s", req.Method, req.URL, resp.Status)
	}
```
The best we can do without updating go-puppetca is checking the content of the error

**Comment by wilfriedroset** [comment]
rebased on master and resolv conflicts

**Pull request merged**